src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -123,7 +123,27 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             }
             Initialize(suggestedCapacity);
 
-            this.UnionWith(collection);
+            var otherAsHashSet = collection as HashSet<T>;
+            if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
+            {
+                _count = otherAsHashSet._count;
+                _lastIndex = otherAsHashSet._lastIndex;
+                _freeList = otherAsHashSet._freeList;
+                _version = otherAsHashSet._version;
+
+                _buckets = new int[otherAsHashSet._buckets.Length];
+                otherAsHashSet._buckets.CopyTo(_buckets, 0);
+
+                _slots = new Slot[otherAsHashSet._slots.Length];
+                otherAsHashSet._slots.CopyTo(_slots, 0);
```

Comment:
Rather than using Initialize with suggestedCapacity, if we're going to take this faster path, we could probably skip the whole Initialize(suggestedCapacity) part and instead just make these:
```C#
_buckets = (int[])otherAsHashSet._buckets.Clone();
_slots = (int[])otherAsHashSet._slots.Clone();
```
In other words, the cast and if check would move up earlier in the method, and all of the stuff around suggestedCapacity and Initialize would move into the else block.  Then we could simplify the if block as above.